### DISCUSSION TITLE
WinMerge add some blank lines when compare two files with line filters

### DISCUSSION THREAD

**Discussion by littlecutebird** [WinMerge/winmerge#988] [Q&A] [answered]
Hello everyone. I have used WinMerge to compare two folders and today I found a very strange problem.
I will use two files in this drive folder to illustrate, you can download from this link: https://drive.google.com/drive/folders/1G96ietGHYlSGKBao3Q3CVlyQnDRC6Hnf?usp=sharing
 
So I compare those two files with WinMerge, apply this line filter: .*\\/\\/.*Token:.\*RID.\*
![Screenshot (10)](https://user-images.githubusercontent.com/33805404/135021219-ed826c6e-e493-44a7-b70e-5c8fe181ae11.png)

This line filter will ignore all differences like this:
![Screenshot (9)](https://user-images.githubusercontent.com/33805404/135021136-1d21ec4b-61e1-4dc3-9266-47f0f09cdc80.png)

But at the end of the file, WinMerge show this difference which I cannot understand:
![Screenshot (11)](https://user-images.githubusercontent.com/33805404/135021283-6d4d41ad-f7b4-402b-99df-faa105435c0e.png)

I have opened two files in VSCode, notepad and haven't seen this blank line. I have also checked the number of lines in two files with command `wc -l` and both files have 2215 lines. Please help me solve this situation because when I compare two folders, there is a huge amount of files getting the difference like the above example.

**Comment by sdottaka** [comment, marked as answer]
> I have opened two files in VSCode, notepad and haven't seen this blank line. 

What you are calling a blank line is a line with a gray background? A line with a gray background is not a real line, but a virtual line that indicates that the corresponding line on one side is missing or has been deleted.

The lines on the right with a gray background is displayed because the Diff algorithm adjusted the left 2207 line and the right 2205 line to be displayed in the same position, since the contents of the left 2207 line "// Token: 0x04005FD RID: 1533" and the right 2205 line are the same.

Unfortunately, WinMerge's Line Filters are applied after the Diff algorithm has adjusted the display position, so you did not get the expected results.

There are plugins called Prediffer plugin that is applied before the comparison. These plugins process the contents of the files before comparison, but the original contents are displayed when they are viewed.

Among the Prediffer plugins, there is a plugin called IgnoreCommentsC (Ignore Comments (C-Familly Languages)). It removes the contents of C++ and C# comment lines before comparing them, and if you apply it, you will get the expected result as shown below.

![image](https://user-images.githubusercontent.com/98126/135091632-4f078676-59d6-4622-9f08-7753e770a14a.png)
![image](https://user-images.githubusercontent.com/98126/135091654-788dfc7b-53a9-4245-97dd-4f31c0921a93.png)

If you want this plugin to be applied automatically when comparing files, select "Plugin" -> "Automatic Prediffer".

![image](https://user-images.githubusercontent.com/98126/135091719-0dd641ff-be9d-4345-b214-be8472134503.png)

**Reply by littlecutebird** [reply]
Thanks! That's helped a lot!
